module/zfs/spa.c

```diff
@@ -4164,6 +4290,126 @@ spa_ld_check_features(spa_t *spa, boolean_t *missing_feat_writep)
 	return (0);
 }
 
+struct load_chain_map_arg {
+	blkptr_t *bp;
+	spa_t *spa;
+	int *error;
+};
+
+static int
+load_chain_map_claim_blk_cb(spa_t *spa, const blkptr_t *bp, void *arg)
+{
+	(void) arg;
+	int error = metaslab_claim(spa, bp,
+	    spa_get_dsl(spa)->dp_tx.tx_open_txg);
+	if (error == ENOENT)
+		error = 0;
+	return (error);
+}
+
+static int
+load_chain_map_claim_lr_cb(spa_t *spa, const lr_t *lrc, void *arg)
+{
+	(void) arg;
+	lr_write_t *lr = (lr_write_t *)lrc;
+	blkptr_t *bp = &lr->lr_blkptr;
+	if (lrc->lrc_txtype != TX_WRITE || BP_IS_HOLE(bp))
+		return (0);
+	int error = metaslab_claim(spa, bp,
+	    spa_get_dsl(spa)->dp_tx.tx_open_txg);
+	if (error == ENOENT)
+		error = 0;
+	return (error);
+}
+
+static void
+load_chain_map_cb(void *arg)
+{
+	struct load_chain_map_arg *lcmca = arg;
+	blkptr_t *bp = lcmca->bp;
+	spa_t *spa = lcmca->spa;
+	int error = zil_parse_raw(spa, bp, load_chain_map_claim_blk_cb,
+	    load_chain_map_claim_lr_cb, NULL);
+	if (error != 0 && error != ECKSUM)
+		*lcmca->error = error;
```

Comment:
there's no locking on `dispatch_error`, I guess we're just assuming this will be an atomic store, and we don't care which error we see?

Reply:
Given that there's a call to `kmem_free` after the write, I think we can be pretty confident that there will be memory barrier-inducing operations before the taskq_wait returns in the calling function. That said, I can change this to be an atomic store and use an atomic read in the caller to fetch the value.